### PR TITLE
Configurable accessors container style

### LINKED ISSUE #19: Suggested non-operator container for accessors / helpers
Right now, helpers are generated into `>` inside the companion object, e.g.:

```scala
@accessable
trait Example {
  val example: Example.Service[Any]
}

object Example {
  trait Service[R] {
    def foo()                 : ZIO[R, Nothing, Unit]
    def bar(v1: Int, v2: Int) : ZIO[R, Nothing, Int]
    def baz(v1: Int)(v2: Int) : ZIO[R, Nothing, String]
  }

  // -- below code is autogenerated -- //
  object > extends Service[R] {
    def foo()                 = ZIO.accessM(_.example.foo)
    def bar(v1: Int, v2: Int) = ZIO.accessM(_.example.bar(v1, v2))
    def baz(v1: Int)(v2: Int) = ZIO.accessM(_.example.baz(v1)(v2))
  }
  // -- end of autogenerated code -- //
}
```

I'd suggest 2 things, first, generate a trait with the helpers, so users have more flexible consumption:

```scala
@accessable
trait Example {
  val example: Example.Service[Any]
}

object Example {
  trait Service[R] {
    def foo()                 : ZIO[R, Nothing, Unit]
    def bar(v1: Int, v2: Int) : ZIO[R, Nothing, Int]
    def baz(v1: Int)(v2: Int) : ZIO[R, Nothing, String]
  }

  // -- below code is autogenerated -- //
  trait Accessors extends Service[R] {
    def foo()                 = ZIO.accessM(_.example.foo)
    def bar(v1: Int, v2: Int) = ZIO.accessM(_.example.bar(v1, v2))
    def baz(v1: Int)(v2: Int) = ZIO.accessM(_.example.baz(v1)(v2))
  }
  // -- end of autogenerated code -- //
}
```

Second, even if `>` remains, I do not think it will be preferred because it has an operator name; so I'd suggest following the ZIO convention to also generate:

```scala
package object example extends Accessors
```

I'm not sure if you can do that, but you could always generate, e.g. `object accessors` inside `Example` companion object, so an import `import Example.accessors._` would bring everything into scope.

### PR DESCRIPTION
**Pull request by ioleo**
Fixes #19

### CODE REVIEW

**Comment by ioleo** [comment]
@jdegoes so I'm generating
* for `@accessible` or `@accessible()` (default case)
```scala
trait Accessors extends Service[Module]
object accessors extends Accessors
```
* for `@accessible(">")` or `@accessible("alias")` (shorthand alias case)
```scala
trait Accessors extends Service[Module]
object > extends Accessors
```
* for `@accessible("trait")` (minimal case)
```scala
trait Accessors extends Service[Module]
```

**Comment by ioleo** [comment]
@mschuwalow Moved delegate & accessible tests to ZIO Test. There is some issue with mockable tests (currently commented) and issues with Scala 2.11 to investigate, but I'm done for today :)

**Review by mschuwalow — COMMENTED** [review]
Oh you migrated to zio-test. Nice!

File: core/shared/src/main/scala/zio/macros/access/Accessible.scala [hunk 1]
```diff
@@ -15,42 +15,70 @@
...
+    @silent("pattern var [^\\s]+ in method unapply is never used")
+    val style: Style = c.prefix.tree match {
+      case Apply(_, args) =>
+        val arg: String = args.collectFirst {
+          case q"$cfg" =>
+            c.eval(c.Expr[String](cfg))
+        }.getOrElse("default")
+
+        arg match {
+          case ">"       => Style.AliasOnly
```
Comment (mschuwalow) [inline core/shared/src/main/scala/zio/macros/access/Accessible.scala]: do we want to make the symbol configurable?

**Review by mschuwalow — APPROVED** [review]
Looks good, but I would prefer configurable accessor symbol

File: core/shared/src/main/scala/zio/macros/access/Accessible.scala (hunk 1 above)
Comment (ioleo) [reply core/shared/src/main/scala/zio/macros/access/Accessible.scala]: I was thinking about it, but wasn't sure if there is such need. Now that you mentioned it I guess it does not hurt to expand the PR :)

**Comment by jdegoes** [comment]
@mschuwalow 

> * for `@accessible` or `@accessible()` (default case)

I like this. Both `accessors` / `Accessors` name and the fact that the existence of the trait makes it easy to mix into custom hierarchies.

> * for `@accessible(">")` or `@accessible("alias")` (shorthand alias case)

Makes sense to make this an alias, an addition to the always-generated `accessors`.

> * for `@accessible("trait")` (minimal case)

Not sure i understand this one.

**Comment by ioleo** [comment]
@jdegoes With `@accessible("trait")` only the `trait Accessors` is generated without the object (`accessors` or `>`). So to use this, the user would have to create an object on his own in the userland code.

For example:

```
@accessible("trait")
trait Foo { val foo: Foo.Service[Any] }
object Foo { trait Service[R] {} }

package object foo extends Foo.Accessors
```

or 

```
@accessible("trait")
trait Foo { val foo: Foo.Service[Any] }
object Foo {
  trait Service[R] {}

  object proxy extends Foo.Accessors
}
```

**Comment by ioleo** [comment]
After @mschuwalow comment im thinking perhaps this would be better:

* `@accessible` or `@accessible()` (default) --> generate only `trait Accessors extends Serivce[Module]`
* `@accessible("valid_scala_identifier")` --> generate trait + object
```scala
@accessible("bar")
trait Foo { val foo: Foo.Service[Any] }
object Foo {
  trait Service[R] { }
  // this is autogenerated
  trait Accessors extends Service[Foo] {}
  object bar extends Accessors
  // end of autogenerated
}
```

Perhaps that would be much cleaner. No arbitrary names. Either a trait only (default), and you get to use it wherever/however you like. Or autogenerated object with configurable name eg: `@accessible("accessors")` or `@accesible(">")`

@mschuwalow @jdegoes WDYT?

**Comment by ioleo** [comment]
@mschuwalow @jdegoes Ready for another review, made the name configurable.

There are still 2 tests (commented out) that I could not get working under 2.11 with ZIO Test. Left some comments to investigate further.

**Comment by ioleo** [comment]
Hmm :thinking: No idea why types mismatch on 2.11. I've checked the generated code output and all methods in `trait Accessors` are typed as `ZIO[Module, ?, ?]`. Given a for-comprehension where all effects require `Module`, why would the whole be seen as `R`?

**Comment by mschuwalow** [comment]
> After @mschuwalow comment im thinking perhaps this would be better:
> 
> * `@accessible` or `@accessible()` (default) --> generate only `trait Accessors extends Serivce[Module]`
> * `@accessible("valid_scala_identifier")` --> generate trait + object
> 
> ```scala
> @accessible("bar")
> trait Foo { val foo: Foo.Service[Any] }
> object Foo {
>   trait Service[R] { }
>   // this is autogenerated
>   trait Accessors extends Service[Foo] {}
>   object bar extends Accessors
>   // end of autogenerated
> }
> ```
> 
> Perhaps that would be much cleaner. No arbitrary names. Either a trait only (default), and you get to use it wherever/however you like. Or autogenerated object with configurable name eg: `@accessible("accessors")` or `@accesible(">")`
> 
> @mschuwalow @jdegoes WDYT?

I like this 👍

**Comment by ioleo** [comment]
@mschuwalow Created seperate issues for the remaining tests migration. Aggregate issue #46

**Pull request merged**